phabricatoremails/render.py

```diff
@@ -0,0 +1,420 @@
+
+    seconds = recipient["timezoneOffset"]  # e.g.: -25200
+    if seconds > 0:
+        hours = floor(seconds / 3600)
+        minutes = floor((seconds - hours * 3600) / 60)
+    else:
+        hours = ceil(seconds / 3600)
+        minutes = ceil((seconds - hours * 3600) / 60)
+    return Recipient(
+        recipient["email"],
+        timezone(timedelta(hours=hours, minutes=minutes)),
+        recipient["isActor"],
+    )
+
+
+def parse_revision(raw_revision):
+    raw_bug = raw_revision.get("bug")
+    bug = Bug(raw_bug["bugId"], raw_bug["name"], raw_bug["link"]) if raw_bug else None
+    return Revision(
+        raw_revision["revisionId"], raw_revision["name"], raw_revision["link"], bug
+    )
+
+
+def parse_secure_revision(raw_revision):
+    raw_bug = raw_revision.get("bug")
+    bug = SecureBug(raw_bug["bugId"], raw_bug["link"]) if raw_bug else None
+    return SecureRevision(raw_revision["revisionId"], raw_revision["link"], bug)
+
+
+def parse_comment_context(context_kind, context):
+    if context_kind == "code":
+        return CodeContext(
+            [
+                DiffLine(
+                    line["lineNumber"], DiffLineType(line["type"]), line["rawContent"]
+                )
+                for line in context["diff"]
+            ]
+        )
+    elif context_kind == "reply":
+        return ReplyContext(
+            context["otherAuthor"],
+            datetime.fromisoformat(context["otherDateUtc"]),
+            context["otherComment"],
+        )
+    else:
+        raise ParseError("Comment context was not code or a reply")
+
+
+def parse_inline_comment(raw_inline):
+    return InlineComment(
+        raw_inline["fileContext"],
+        raw_inline["link"],
+        raw_inline["text"],
+        parse_comment_context(raw_inline["contextKind"], raw_inline["context"]),
+    )
+
+
+def parse_affected_file(raw_file):
+    return AffectedFile(raw_file["path"], AffectedFileChange(raw_file["change"]))
+
+
+def parse_reviewer(raw_reviewer):
+    return Reviewer(
+        raw_reviewer["name"],
+        raw_reviewer["isActionable"],
+        ReviewerStatus(raw_reviewer["status"]),
+        parse_recipient(raw_reviewer.get("recipient")),
+    )
+
+
+def parse_metadata_edited_reviewer(raw_reviewer):
+    return MetadataEditedReviewer(
+        raw_reviewer["name"],
+        raw_reviewer["isActionable"],
+        ReviewerStatus(raw_reviewer["status"]),
+        ExistenceChange(raw_reviewer["metadataChange"]),
+        parse_recipient(raw_reviewer.get("recipient")),
+    )
+
+
+def render(logger: Logger, stats: StatsClient, events: List, db: Session, css_inline):
+    outgoing_emails = []
+    template_store = TemplateStore()
+    for event in events:
+        try:
+            is_secure = event["isSecure"]
+            kind = event["eventKind"]
+            timestamp = event["timestamp"]
+            body = event["body"]
+            batch = MailBatch(css_inline)
+
+            if kind == "revision-accepted":
```

Comment:
Hmm, these strings are used in a single location to parse a data structure.
I can understand documenting the functionality here: "we check `kind` to identify the data structure, then parse accordingly".

However, what is the benefit of defining these strings elsewhere? If you put them into a `const` or `Enum` somewhere else, I don't understand how that improves maintainability. Rather, I understand that that would make it harder to understand, since readers of the function would have to jump to the constant/enum to understand what the value is supposed to be.

Reply:
They don't have to be defined far from this method (and in fact, they should be defined in this module.) Putting all the possible values in a single place will make it easier to see what possible expected values are, and can even be used to validate early on whether the value is in an expected set of values.

When you write unit tests for these methods, you'll want to import those values and use them in your tests, rather than rewriting the strings in each test. They can also be used in mocks, etc...

>  I don't understand how that improves maintainability. Rather, that that would make it harder to understand, since readers of the function would have to jump to the constant/enum to understand what the value is supposed to be.

The keys you pick for these values should be easy to understand without a user having to go back and forth too many times. Also, if these values are fixed for a specific reason (e.g. to be compatible with an external service), the exact string value should not matter for the purposes of maintaining this application. If the values change in the third party service, it would in fact be easier to change them in the enum rather than in each if statement, etc...

Reply:
I've added a `KIND` property to each event dataclass :+1: